### DISCUSSION TITLE
[Bug] DockerComposeContainer.stop() does not set env in versions 1.19.X

### DISCUSSION THREAD

**Discussion by GStepien** [testcontainers/testcontainers-java#7965] [General]
**Module**
Core

**Testcontainers version**
1.19.3

**Using the latest Testcontainers version?**
Yes

Hi, 
after updating our testcontainers dependency from 1.18.3 to 1.19.3, we get a number of errors in the testing stage of our buildpipeline when our testmethod calls stop() on a DockerComposeContainer instance. The error messages indicate that the environment variables referenced in our docker compose YAML file do not get dereferenced in the respective `docke-compose ... down` call which leads to the mentioned errors. We did not have this problem in pre 1.19.0 versions and have, for now, indeed downgraded our testcontainers dependency to 1.18.3.

After some digging in the testcontainers code, I think the problem is this line in the stop() method:
https://github.com/testcontainers/testcontainers-java/blob/10ecdcee8eeb5ec3d9bc841c642b25f3823cb4e8/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java#L175
The runWithCompose method is called without the env field leading to the aforementioned problems. I think this line should be:
`this.composeDelegate.runWithCompose(this.localCompose, cmd, this.env);`
If I am not mistanken, this should result in a behaviour equivalent to pre-1.19.X verions.

Any thoughts on this?

Best regards

**Comment by eddumelendez** [comment]
Hi, can you share a project that reproduces the issue, please?

**Reply by GStepien** [reply]
Hi, 
unfortunately this is a proprietary company project which I cannot share here. I hope that my code analysis is convincing enough:
In version 1.18.3 we see, that the `stop()` method calls `runWithCompose(cmd)`:
https://github.com/testcontainers/testcontainers-java/blob/b0902a95e7451e27e13815c48fe65314351d40e2/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java#L381
which, in turn, invokes the docker compose command WITH the environment variables set:
https://github.com/testcontainers/testcontainers-java/blob/b0902a95e7451e27e13815c48fe65314351d40e2/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java#L339

This behavour changed in versions >= 1.19.0:
Now the `stop()` method calls `runWithCompose(this.localCompose, cmd)` on a compose delegate:
https://github.com/testcontainers/testcontainers-java/blob/10ecdcee8eeb5ec3d9bc841c642b25f3823cb4e8/core/src/main/java/org/testcontainers/containers/DockerComposeContainer.java#L175
which, in turn, calls `runWithCompose(boolean localCompose, String cmd, Map<String, String> env)` with an EMPTY environment variable map here:
https://github.com/testcontainers/testcontainers-java/blob/10ecdcee8eeb5ec3d9bc841c642b25f3823cb4e8/core/src/main/java/org/testcontainers/containers/ComposeDelegate.java#L240

This results in environment variables from the YAML File not being substituted when the underlying `docker compose ... down` command is executed. I assume this apparent change in behaviour was not intentional or was it? ;) I think the bugfix I proposed above should fix the issue (i.e., provide the env variables in the `stop()` method when calling runWithCompose on the delegate).

Best regards

**Reply by eddumelendez** [reply]
Hi, there is no need to share the same project but a little one that reproduces the issue. I have also look into the code and see there is a difference between previous a new implementation but there is no test that can produce anything that you described. So, having your project would help to capture it in a test.
